### PR TITLE
postgresql_user: fix bugs related to 'expires' option

### PR DESCRIPTION
**Pull request by pilou-**
##### SUMMARY
Add tests and fix various errors related to the `expires` option. The last 4 commits improve readability.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
postgresql_user

##### ANSIBLE VERSION
```
ansible-playbook 2.4.0 (devel 8ec4882ba0)
```

##### ADDITIONAL INFORMATION

Locally tested using:
```
ANSIBLE_REMOTE_USER=root ANSIBLE_ROLES_PATH=test/integration/targets ansible-playbook -i 'testhost,' -e ansible_host=10.0.1.129 --tags=test_postgresql,test_postgresql_db,test_postgresql_privs,test_postgresql_user,needs_privileged test/integration/destructive.yml
```

Related pull-requests:
* #23613 PostgreSQL integration test suite: really generate locales on Debian
* #23631 PostgreSQL Integration test suite: improvements
* #23686 postgresql_user: use standard module hashlib instead of passlib

### CODE REVIEW

File: lib/ansible/modules/database/postgresql/postgresql_user.py [hunk 1]
```diff
@@ -354,9 +354,7 @@ def user_alter(cursor, module, user, password, role_attr_flags, encrypted, expir
         new_role_attrs = cursor.fetchone()
 
         # Detect any differences between current_ and new_role_attrs.
-        for i in range(len(current_role_attrs)):
-            if current_role_attrs[i] != new_role_attrs[i]:
-                changed = True
+        changed = current_role_attrs != new_role_attrs
```
Comment (mscherer) [inline lib/ansible/modules/database/postgresql/postgresql_user.py:406]: So , would it work for all psycopg2 version ?

**Review by mscherer — COMMENTED** [review]
All look goods, just one question.

File: lib/ansible/modules/database/postgresql/postgresql_user.py (hunk 1 above)
Comment (pilou-) [reply lib/ansible/modules/database/postgresql/postgresql_user.py:406]: `psycopg2.extras.DictRow` was already here in [1.99.12](https://github.com/psycopg/psycopg2/blob/1_99_12/lib/extras.py#L61) (which was released in march 2005).

**Comment by drrtuy** [comment]
There is one more problem with "expires" code: there is no column rol_valid_until in a [pg_authid relation](https://pastebin.com/5cWGwyHx). So [this code](https://github.com/pilou-/ansible/blob/4359c8512c4f45f4d18994101486b8e84184216a/lib/ansible/modules/database/postgresql/postgresql_user.py#L323) always compare expires with a None. It should be changed.

**Comment by pilou-** [comment]
@drrtuy thanks for pointing that. I will update this pull request after the merge of #23686.

**Comment by matburt** [comment]
Looks like some test failures, would like to see those pass before merge

**Comment by pilou-** [comment]
Me too :) Locally tests don't fail, I will check the errors as soon as shippable will be available.

**Comment by pilou-** [comment]
@matburt @drrtuy @mscherer : ready_for_review

File: lib/ansible/modules/database/postgresql/postgresql_user.py [hunk 2]
```diff
@@ -207,11 +207,11 @@
     password: NULL
 '''
 
+from distutils.version import StrictVersion
 from hashlib import md5
 import itertools
 import re
 
-from distutils.version import StrictVersion
```
Comment (abadger) [inline lib/ansible/modules/database/postgresql/postgresql_user.py]: I'd move the hashlib import down here... imports are usually
```
import stdlib_foo
from stdlib_foo import bar

import thridparty_foo
from thirdparty_foo import bar

import project_foo
from project_foo import bar
```

Comment (pilou-) [reply lib/ansible/modules/database/postgresql/postgresql_user.py]: `hashlib` module belongs to the Python standard library ([2.7](https://docs.python.org/2.7/library/hashlib.html), [3.6.1](https://docs.python.org/3/library/hashlib.html)).

Comment (sivel) [reply lib/ansible/modules/database/postgresql/postgresql_user.py]: Based on @abadger's comment, due to it being `from hashlib import md5` it should go below `import re`

Comment (pilou-) [reply lib/ansible/modules/database/postgresql/postgresql_user.py]: oh right, I misunderstood! I fixed that and other PEP 8 related problems.

**Comment by mscherer** [comment]
shipit

**Comment by abadger** [comment]
Merged to devel.  Doesn't cherrypick cleanly and I'm going on a one week vacation followed by ansiblefest London so I didn't want to backport it.  If you think it should be backported, feel free to ask one of the other devs if they think it should be.

**Comment by pilou-** [comment]
This pull-request contains 4 one-liner fixes and many commits related to the tests. Without the bugfixes, `expires` parameter isn't usable.

Are there any recommendations/policies about backport of bugfixes ?

I have created a [branch](https://github.com/pilou-/ansible/tree/postgresql_user_fix_valid_until_2.3) based on stable-2.3 with the 4 one-liners cherry picked. I have referenced the squashed commit, should I use commit id before the squash instead ?

**Pull request merged**